### PR TITLE
[ENH] KSpectralCentroid clusterer

### PR DESCRIPTION
**Pull request by chrisholder**
#### Reference Issues/PRs
Relates to issue #2129
This PR depends on #2139

#### What does this implement/fix? Explain your changes.
This PR implements the KSpectralCentroid (k-SC) clusterer. k-SC is similar to k-means but uses a bespoke averaging and distance measure that is invariant to scale and shift.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [x] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/clustering/averaging/_shift_scale_invariant_averaging.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                initial_center, X[i], max_shift
+            )
+            optimal_shifts[i] = curr_shift
+
+    if optimal_shifts.shape[0] == 0:
+        return np.zeros((n_dims, n_timepoints))
+
+    normalized_shifts = optimal_shifts / np.sqrt(
+        np.sum(optimal_shifts**2, axis=2)
+    ).reshape(n_instances, n_dims, 1)
```
Comment (baraline) [inline aeon/clustering/averaging/_shift_scale_invariant_averaging.py:38]: I'm missing the goal of the reshape here `optimal_shifts` is supposed to be 3D right ?

File: aeon/clustering/averaging/_shift_scale_invariant_averaging.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            normalized_shifts[:, d, :].T, normalized_shifts[:, d, :]
+        ) - n_instances * np.eye(n_timepoints)
+
+    new_center = np.zeros((n_dims, n_timepoints))
+    for d in range(n_dims):
+        eigvals, eigvecs = np.linalg.eig(M[d])
+        new_center[d] = np.real(eigvecs[:, np.argmax(eigvals)])
+
+        if np.sum(new_center[d]) < 0:
+            new_center[d] = -new_center[d]
```
Comment (baraline) [inline aeon/clustering/averaging/_shift_scale_invariant_averaging.py:52]: Do you really need the matrix M and the two loops here ? It looks like there is no interactions other than storing the result of the dot product to be used in the second loop.

**Review by baraline — COMMENTED** [review]
Minor comments mostly for my understanding, rest looks good

File: aeon/clustering/averaging/_shift_scale_invariant_averaging.py (hunk 1 above)
Comment (chrisholder) [reply aeon/clustering/averaging/_shift_scale_invariant_averaging.py:38]: The reason is due to how np.sqrt and np.sum change the shapes of the input.

For example:

``` 
   # Assume optimal_shift is of shape (12, 1, 10) meaning there the optimal 
   # shift for 12 time series instances has been computed

    # If we sum this array
    sum_of = np.sum(optimal_shifts**2, axis=2)
   # sum_of will be  an array of shape (12, 1). This means for each shift their total sum

   # If we sqrt then sqrt to normalise
   sqrt_of = np.sqrt(sum_of)
   # sqrt_of will be an array of shape (12, 1) meaning there are 12 different 
   # sqrts for each shift


  # If we then want to divide optimal shifts which is of shape (12, 1, 10)
  # by sqrt_of which is of shape (12, 1)
  # We first have to reshape sqrt_of so it is of shape (12, 1, 1) so it 
  # matches the shape of the optimal shifts
  reshape_sqrt = sqrt_of.reshape(n_instances, n_dims, 1)
  
  normalized_shifts = optimal_shifts / reshape_sqrt
```

File: aeon/clustering/averaging/_shift_scale_invariant_averaging.py (hunk 2 above)
Comment (chrisholder) [reply aeon/clustering/averaging/_shift_scale_invariant_averaging.py:52]: Yes this is a very good point I have made it into a single loop now.

File: aeon/clustering/averaging/_shift_scale_invariant_averaging.py (hunk 1 above)
Comment (baraline) [reply aeon/clustering/averaging/_shift_scale_invariant_averaging.py]: Missed the sum function my bad !

**Review by baraline — APPROVED** [review]
Lgtm !

**Pull request merged**